pkg/sql/catalog/tabledesc/constraint.go

```diff
@@ -0,0 +1,238 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tabledesc
+
+import (
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+)
+
+// constraintCache contains precomputed slices of constraints, categorized by kind and validity.
+// A constraint is considered
+// - active if its validity is VALIDATED;
+// - inactive if its validity is VALIDATING or UNVALIDATED;
+// - dropping if its validity is DROPPING;
+type constraintCache struct {
+	all                  []catalog.Constraint
+	allActiveAndInactive []catalog.Constraint
+	allActive            []catalog.Constraint
+
+	allChecks                  []catalog.Constraint
+	allActiveAndInactiveChecks []catalog.Constraint
+	allActiveChecks            []catalog.Constraint
+
+	allNotNulls                  []catalog.Constraint
+	allActiveAndInactiveNotNulls []catalog.Constraint
+	allActiveNotNulls            []catalog.Constraint
+
+	allFKs                  []catalog.Constraint
+	allActiveAndInactiveFKs []catalog.Constraint
+	allActiveFKs            []catalog.Constraint
+
+	allUniqueWithoutIndexs                  []catalog.Constraint
```

Comment:
nit: `s/allUniqueWithoutIndexs/allUniqueWithoutIndexes`; same for 2 lines below.

Reply:
done